### PR TITLE
4 create spatial memory game

### LINKED ISSUE #4: Create spatial memory game
## Feature Description
Create a new component that allows the user to play a spatial memory game. 

## User Story
"As a user I would like to be tested on my spatial memory ability."

## MVP
A new react component that displays a grid of intractable squares. The game will flash with a pattern and the user can match the pattern by clicking on the squares. The user will be notified if the input matched the pattern or not.

### PR DESCRIPTION
**Pull request by JackHH7297**
## Issue
Create 2 spatial memory game:

Game 1: Card matching game
Game 2: Pattern memory game

### Card matching game:
Instruction Page:
image

Game interface:
image

End game:
image

Cards locations are revealed to users at the beginning of the game temporarily. Users should be able to click on a card to flip it, if it matches user's first selection then the pair of cards will stay flipped. Game ends when users run out of lives or time.

### Pattern memory game:
Game interface:
image

Users are shown a pattern of X number of blocks for a short time, then users need to click on blocks one by one to reproduce the pattern. When users click on a block that is a part of the pattern then it will change to a green color, otherwise grey.Game will progressively become more diffcult by increasing number of blocks in the pattern. Game ends when users run out of lives or time.

Reviwed by
Damon

### CODE REVIEW

File: client/src/components/MatchingGame/GameOver.js
Comment (DamonGreenhalgh) [inline client/src/components/MatchingGame/GameOver.js]: You don't have to define new components for these elements. Something like,
```
{
    default || victory ?
    <div className='gameover-text-block'>
        <h2 className='gameover-text'>{victory ? 'Good job, You won!' : 'Game Over'}</h2>
    </div>
}
```
would work and you can put it where the `GameOver` component was.

File: client/src/components/MatchingGame/LivesCounter.js
Comment (DamonGreenhalgh) [inline client/src/components/MatchingGame/LivesCounter.js]: Since these components have little to no logic tied to them it would be best to just have them within the `MatchingGame` component. Something like,
```
{
    !onLaunch ?
    <div className='lives-div'>
        <h2 className='livesNum'>Lives:</h2>
        {[...Array(numLives)].map((e, i) => <span className="heart" key={i} />)}
    </div>
}
```
in place of `LivesCounter`.

File: client/src/components/MatchingGame/MatchingGame.js
Comment (DamonGreenhalgh) [inline client/src/components/MatchingGame/MatchingGame.js]: In general, if there is not much logic tied to a component you can just put it within the parent component with some conditional rendering.

File: client/src/components/MatchingGame/SingleCard.js
Comment (DamonGreenhalgh) [inline client/src/components/MatchingGame/SingleCard.js]: You can put the function within the prop, 

`onClick={() => !disabled ? handleChoice(card) : null}`

File: client/src/styles/MatchingGameStyles/GameOver.css
Comment (DamonGreenhalgh) [inline client/src/styles/MatchingGameStyles/GameOver.css]: Try to avoid using camelCase for CSS classes. Look into [BEM](http://getbem.com/naming/) for naming conventions.
`.gameOver-text-block` to `gameover__text-block`

**Review by DamonGreenhalgh — CHANGES_REQUESTED** [review]
Good work on the card matching game. Just needs few changes for better readability with conditional rendering and CSS naming conventions. As it looks like you have integrated @Vladd242's utility components we'll wait on those to merge to `dev` before merging this branch.

File: client/src/components/MatchingGame/GameOver.js
Comment (JackHH7297) [reply client/src/components/MatchingGame/GameOver.js]: I see, I did it because I think I am going to be reusing this component for the pattern game.

File: client/src/components/MatchingGame/LivesCounter.js
Comment (JackHH7297) [reply client/src/components/MatchingGame/LivesCounter.js]: Same reason as game over component, thought I will reuse these for the other game. I guess I will just repeat these separately.

File: client/src/styles/MatchingGameStyles/GameOver.css
Comment (JackHH7297) [reply client/src/styles/MatchingGameStyles/GameOver.css]: Yep, I will look into that

File: client/src/components/MatchingGame/MatchingGame.js
Comment (JackHH7297) [reply client/src/components/MatchingGame/MatchingGame.js]: Does it mean I should just put this block within the return statement?

File: client/src/components/MatchingGame/SingleCard.js
Comment (JackHH7297) [reply client/src/components/MatchingGame/SingleCard.js]: Will change. I separated them to help me understand the code better.

File: client/src/components/MatchingGame/MatchingGame.js
Comment (DamonGreenhalgh) [reply client/src/components/MatchingGame/MatchingGame.js]: Yeah, you can put both the `DisplayTime` and `Instructions` within the `MatchingGame` return.

**Pull request merged**
